### PR TITLE
More cleanup

### PR DESCRIPTION
**Pull request by boaks**
More handshake cleanup.
Small fixes for connection.
Check server_hello parameter.
 
Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by boaks** [comment]
@sbernard31 

May be you can have a fast try with this and the client of issue #1034

File: scandium-core/src/test/java/org/eclipse/californium/scandium/DTLSConnectorTest.java [hunk 1]
```diff
@@ -527,7 +530,9 @@ public void testAcceptClientHelloAfterIncompleteHandshake() throws Exception {
 		client = new DTLSConnector(clientConfig, clientConnectionStore);
 
 		// THEN assert that the handshake succeeds and a session is established
-		givenAnEstablishedSession();
+		// sometime the retransmission of the previous incomplete handshake
+		// fails the new handshake. Therefore a retry is used.
+		givenAnEstablishedSessionWithRetry();
```
Comment (sbernard31) [inline scandium-core/src/test/java/org/eclipse/californium/scandium/DTLSConnectorTest.java]: > sometime the retransmission of the previous incomplete handshake fails the new handshake.

I'm curious : how does it fail ?

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ResumingClientHandshaker.java [hunk 2]
```diff
@@ -159,6 +159,7 @@ protected void doProcessMessage(DTLSMessage message) throws HandshakeException,
 					// Server refuse to resume the session, go for a full handshake
 					fullHandshake  = true;
 					super.receivedServerHello(serverHello);
+					incrementNextReceiveSeq();
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ResumingClientHandshaker.java]: Good catch !
Any reason to not just remove the `return;` ?

File: scandium-core/src/test/java/org/eclipse/californium/scandium/dtls/AdversaryClientHandshaker.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+ *                                                    trustStore := null, disable x.509
+ *                                                    trustStore := [], enable x.509, trust all
+ *    Achim Kraus (Bosch Software Innovations GmbH) - fix usage of literal address
+ *                                                    with enabled sni
+ *    Vikram (University of Rostock) - added ECDHE_PSK mode
+ *    Achim Kraus (Bosch Software Innovations GmbH) - add handshake parameter available to
+ *                                                    process reordered handshake messages
+ *    Achim Kraus (Bosch Software Innovations GmbH) - add dtls flight number
+ *    Achim Kraus (Bosch Software Innovations GmbH) - redesign DTLSFlight and RecordLayer
+ ******************************************************************************/
```
Comment (sbernard31) [inline scandium-core/src/test/java/org/eclipse/californium/scandium/dtls/AdversaryClientHandshaker.java]: is it intended to keep all this contributors in the header of this new class ?

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java [hunk 4]
```diff
@@ -1103,7 +1105,15 @@ private DTLSMessage handshakeMessageFromByteArray(byte[] decryptedMessage) throw
...
-		return HandshakeMessage.fromByteArray(decryptedMessage, parameter, getPeerAddress());
+		HandshakeMessage handshakeMessage = HandshakeMessage.fromByteArray(decryptedMessage, parameter, getPeerAddress());
+		if (handshakeMessage.getMessageType() == HandshakeType.FINISHED && epoch == 0) {
+			LOGGER.debug("FINISH with epoch 0 from peer [{}]!",
+					peerAddress);
+			AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.UNEXPECTED_MESSAGE,
+					peerAddress);
+			throw new HandshakeException("FINISH with epoch 0!", alert);
+		}
+		return handshakeMessage;
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java]: It sounds like Handshaker logic so I would move it in an `Handshaker` :thinking:

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Connection.java [hunk 5]
```diff
@@ -519,8 +519,12 @@ public String toString() {
 		if (peerAddress != null) {
 			builder.append(", ").append(peerAddress);
 			if (getOngoingHandshake() != null) {
-				String id = getOngoingHandshake().getSession().getSessionIdentifier().getAsString().substring(0,  8);
-				builder.append(", ongoing handshake ").append(id);
+				builder.append(", ongoing handshake ");
+				SessionId id = getOngoingHandshake().getSession().getSessionIdentifier();
+				if (id != null) {
+					// during handshake this may by not already set
+					builder.append(id.getAsString().substring(0,  8));
```
[the remark this answers is not in the document]
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Connection.java]: To have shorter logging messages.

File: scandium-core/src/test/java/org/eclipse/californium/scandium/DTLSConnectorTest.java (hunk 1 above)
Comment (boaks) [reply scandium-core/src/test/java/org/eclipse/californium/scandium/DTLSConnectorTest.java]: First handshake is in progress. Potential, the last pending flight is retransmitted.
Second handshake sends the 2. client_hello. If now the server_hello from the first is received and processed, the handshake fails. There is a very small window, when the last pending flight is resend and the new client_hello is received on the server side. You may use the repeated test runner to see, that without the retry, the test fails rarely.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ResumingClientHandshaker.java (hunk 2 above)
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ResumingClientHandshaker.java]: With the return, it's clear the the control flow of the full handshake is used. 
Even if the current implementation would not differ.

File: scandium-core/src/test/java/org/eclipse/californium/scandium/dtls/AdversaryClientHandshaker.java (hunk 3 above)
Comment (boaks) [reply scandium-core/src/test/java/org/eclipse/californium/scandium/dtls/AdversaryClientHandshaker.java]: Uups, copy&paste

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java (hunk 4 above)
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java]: My idea was, that a FINISH epoch 0 is a invalid record. But I can move it into the handshaker.

**Comment by boaks** [comment]
Remove the wrong header license.
Move FINISH epoch 0 to Handshaker

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Connection.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sbernard31) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Connection.java]: Personally I would keep the complete sessionId but it's OK like this too.

File: scandium-core/src/test/java/org/eclipse/californium/scandium/DTLSConnectorTest.java (hunk 1 above)
Comment (sbernard31) [reply scandium-core/src/test/java/org/eclipse/californium/scandium/DTLSConnectorTest.java]: Ok I see.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ResumingClientHandshaker.java (hunk 2 above)
Comment (sbernard31) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ResumingClientHandshaker.java]: Ok but here we just want to increment and log we receive a message and this is was is done later :)

**Comment by sbernard31** [comment]
LGTM.

By the way, I don't have any devices available so I can not really test it. But I would let you know if I find a way to get feedback about that.

**Pull request merged**